### PR TITLE
feat: Add `reboot-required` annotation

### PR DESCRIPTION
**Pull request by timo-42**
#702

### CODE REVIEW

File: cmd/kured/main.go [hunk 1]
```diff
@@ -652,11 +670,9 @@ func rebootAsRequired(nodeID string, rebootCommand []string, sentinelCommand []s
 			// And finally (3) if it has that annotation, to delete it.
 			// This indicates to other node tools running on the cluster that this node may be a candidate for maintenance
 			if annotateNodes && !rebootRequired(sentinelCommand) {
-				if _, ok := node.Annotations[KuredRebootInProgressAnnotation]; ok {
-					err := deleteNodeAnnotation(client, nodeID, KuredRebootInProgressAnnotation)
-					if err != nil {
-						continue
-					}
+				err := deleteNodeAnnotation(client, nodeID, KuredRebootInProgressAnnotation)
```
Comment (ckotzbauer) [inline cmd/kured/main.go:673]: Why the if was removed? It would avoid a call to the apiserver.

**Review by ckotzbauer — CHANGES_REQUESTED** [review]
Thanks for your PR. I added one smaller comment above.
I think it would make sense to only add/remove this annotation when the `--annotate-nodes` flag is set. This would streamline the new annotation with the others.

File: cmd/kured/main.go (hunk 1 above)
Comment (timo-42) [reply cmd/kured/main.go]: Readded the check.

**Comment by timo-42** [comment]
> Thanks for your PR. I added one smaller comment above. I think it would make sense to only add/remove this annotation when the `--annotate-nodes` flag is set. This would streamline the new annotation with the others.

Put the new annotation behind the annotation toggle.

**Comment by timo-42** [comment]
@ckotzbauer I pushed the changes you requested.

**Comment by timo-42** [comment]
@ckotzbauer @dholbach Can you please have a look?

**Comment by ckotzbauer** [comment]
@timo-42 I have not forgotten, but I am a bit busy at the moment.

**Comment by timo-42** [comment]
You're right. I did not consider checking the state outside the window! Moved the code to a separate goroutine.
The question is: Should we create a separate goroutine for prometheus metric and the annotations or should we merge it?

**Comment by vitality411** [comment]
I'm not a developer per se, but I think it would have advantages for the memory footprint to have only one goroutine. Especially since they partially execute the same commands.

**Comment by timo-42** [comment]
Moved code into metric goroutine.

**Comment by timo-42** [comment]
@vitality411 @ckotzbauer Is there anything more I should do?

**Comment by vitality411** [comment]
Hi @timo-42, from my side it looks good.

**Comment by jackfrancis** [comment]
@timo-42 Thank you for your patience during the review process, I'm back online and able to do this responsively going forward.

As I understand it, what you'd like to accomplish here is to not schedule certain workloads on nodes that are running kured, because those nodes will responsively reboot, and cordon/drain is not sufficient to gracefully evict and reschedule those workloads. Is that correct?

If so, I wonder if this issue would be a more elegant solution: https://github.com/kubereboot/kured/issues/737

My observation: there is a problem with using a node annotation via the kured daemonset to identify "I am running the kured daemonset", because if for any reason that daemonset is removed manually, and/or the appropriate node taints are added so that it is no longer running kured, that annotation will have to be manually removed (and inevitably this will not happen 100% of the time, and so that annotation will not be 100% accurate in reflecting the state of that node).

**Comment by timo-42** [comment]
> As I understand it, what you'd like to accomplish here is to not schedule certain workloads on nodes that are running kured, because those nodes will responsively reboot, and cordon/drain is not sufficient to gracefully evict and reschedule those workloads. Is that correct?

Yes. We want to not schedule new workload on it and want to be notified that a node needs a reboot, so we can start draining the pods in their respective maintenance windows.

> If so, I wonder if this issue would be a more elegant solution: #737

If I understand this correctly, the proposal is to `run an arbitrary kubectl command`  before draining. This is already to late in the reboot process for us. We need days to drain a node, therefore we need the annotation as soon as possible and before the kured acquires the cluster lock and runs the reboot command.

> My observation: there is a problem with using a node annotation via the kured daemonset to identify "I am running the kured daemonset", because if for any reason that daemonset is removed manually, and/or the appropriate node taints are added so that it is no longer running kured, that annotation will have to be manually removed (and inevitably this will not happen 100% of the time, and so that annotation will not be 100% accurate in reflecting the state of that node).

If the concern is cleanup, than we could add a [pre delete hook](https://helm.sh/docs/topics/charts_hooks/#the-available-hooks) to delete all dangling annotations.

**Comment by jackfrancis** [comment]
@timo-42 Thanks for the explanation.

Here's my thinking: by putting the annotation on the node as soon as kured launches its daemonset runtime you are essentially declaring: don't schedule any non-evictable workloads on this node because I *may* be rebooted at some point in the future. And then you would begin the process of evicting existing workloads, which may take days. But it's not clear to me how having this advance notice will actually guarantee that the underlying host OS will not put a `/var/run/reboot-required` file before that days-long eviction completes. And so you are still at risk of the node rebooting during the workload eviction process.

It would seem to me that the workload requirements you're outlining would forbid kured from running at all on the nodes that are running those workloads. What is the downside of simply tainting such nodes and not having kured run on it. I'm not sure that kured is as it currently is a viable solution for these operational requirements.

@dholbach @ckotzbauer @evrardjp any other thoughts?

**Comment by timo-42** [comment]
@jackfrancis 
> It would seem to me that the workload requirements you're outlining would forbid kured from running at all on the nodes that are running those workloads. What is the downside of simply tainting such nodes and not having kured run on it. I'm not sure that kured is as it currently is a viable solution for these operational requirements.

This is easy :) - I use `--blocking-pod-selector` to prevent rebooting nodes which have these kinds of workloads.

**Comment by jackfrancis** [comment]
@timo-42 Thanks for clarifying, that makes sense!

In this case this would be my suggestion (curious for others' thoughts as well):

Rather than add a third annotation to a node to describe kured reboot state, let's simply re-order the steps in the flow where we annotate `"weave.works/kured-reboot-in-progress"`. At present, we short-circuit the flow if a matching blocking pod selector is found, but we can short-circuit due to that critera _later_ in the flow. For example:

If a reboot is required:
1. Add the `"weave.works/kured-reboot-in-progress"` annotation
2. Add the "prefer no schedule" taint to the node
3. If reboot is blocked due to matching pod selector, continue
  a. Else, cordon/drain and reboot node

The above should solve for your use-case. You may use the `"weave.works/kured-reboot-in-progress"` annotation in your own workflows to prevent new, non-interruptible workloads from landing on this node (or you may be able to rely upon the "prefer no schedule" taint, but that taint is not strict, so FYI). Only when existing, non-interruptible workloads have gracefully exited, and the pod selector finds no match, will the node actually cordon/drain + reboot.

@ckotzbauer @dholbach @evrardjp if they have any criticisms to the above

**Comment by ckotzbauer** [comment]
I like the idea @jackfrancis. I'm not fully aware of the current implementation of `"weave.works/kured-reboot-in-progress"`: Would this be a breaking change?

**Comment by jackfrancis** [comment]
@ckotzbauer In my view it wouldn't be a breaking change. The original implementation of `"weave.works/kured-reboot-in-progress"` was meant to annotate the node *as soon as we know it will eventually be rebooted*. One particular condition that can delay a reboot from happening right away is the inability to acquire the exclusive node reboot lock. But we still want to annotate the node because in the next loop we may require that lock, and thus we still don't want applications that are consuming that annotation to consider that node as a viable node for scheduling new workloads.

It seems that the blocking pod selector criteria was overlooked when implementing the `"weave.works/kured-reboot-in-progress"` annotation, but that criteria is functionally equivalent to the exclusive node reboot lock: in the next loop there may be no more pods that match the blocking pod criteria, and so we are in a similar situation where we know that the node will eventually be rebooted.

tl;dr I think there's a bug in the implementation of the `"weave.works/kured-reboot-in-progress"` node annotation, and if we fix that bug, then @timo-42 and other folks can use _it_ to build conditional scheduling flows around.

@timo-42 does that make sense? I'd be happy to fix that bug myself.

**Comment by timo-42** [comment]
Closing since #749 got merged.
